crates/pacdef_core/src/backend/actual/rustup.rs

```diff
@@ -0,0 +1,208 @@
+            let toolchain = iter
+                .next()
+                .expect("No toolchain name provided for component");
+            if repo == "component" && !toolchains_rem.contains(&toolchain) {
+                let mut cmd = Command::new(self.get_binary());
+                cmd.args(self.get_remove_switches(repo)).arg(toolchain).arg(
+                    iter.next()
+                        .expect(format!("Component name not provided for {}", p.name).as_str()),
+                );
+                result = cmd.status().context("Removing toolchain {p}");
+                if !result.as_ref().is_ok_and(|exit| exit.success()) {
+                    return result;
+                }
+            }
+        }
+        result
+    }
+}
+
+impl Rustup {
+    pub(crate) fn new() -> Self {
+        Self {
+            packages: HashSet::new(),
+        }
+    }
+
+    fn get_install_switches(&self, repotype: &str) -> Switches {
+        match repotype {
+            "toolchain" => &["toolchain", "install"],
+            "component" => &["component", "add", "--toolchain"],
+            _ => panic!("No such type managed by rust"),
+        }
+    }
+
+    fn get_remove_switches(&self, repotype: &str) -> Switches {
+        match repotype {
+            "toolchain" => &["toolchain", "uninstall"],
+            "component" => &["component", "remove", "--toolchain"],
+            _ => panic!("No such type managed by rust"),
+        }
+    }
+
+    fn get_info_switches(&self, repotype: &str) -> Switches {
+        match repotype {
+            "toolchain" => &["toolchain", "list"],
+            "component" => &["component", "list", "--installed", "--toolchain"],
+            _ => panic!("No such type managed by rust"),
+        }
+    }
+
+    fn run_component_command(
+        &self,
+        args: &[&str],
+        toolchains: &mut Vec<String>,
+    ) -> Result<Vec<String>, anyhow::Error> {
+        let mut val = Vec::new();
+        for toolchain in toolchains {
+            let mut cmd = Command::new(self.get_binary());
+            cmd.args(args).arg(&toolchain);
+            let output = String::from_utf8(cmd.output()?.stdout)?;
+            for i in output.lines() {
+                let mut it = i.splitn(3, "-");
+                let component = it.next().expect("Component name is empty!");
+                match component {
+                    "cargo" | "rustfmt" | "clippy" | "miri" | "rls" | "rustc" => {
+                        val.push([toolchain, component].join("/"));
+                    }
+                    _ => {
+                        let component = [
+                            component,
+                            it.next().expect("No such component is managed by rustup"),
+                        ]
+                        .join("-");
+                        val.push([toolchain, component.as_str()].join("/"));
+                    }
+                }
+            }
+        }
+        Ok(val)
+    }
```

Comment:
The first two points are taken care of. I'll tackle the last one.

Reply:
As mentioned in another review, I'd like to know when to panic and when not to.

Reply:
See above

Reply:
Reposting it here, because Github seems to have swallowed it. 



In Rust, panics are generally for programming errors and unrecoverable errors. Result types (and their Err variant) are used for user input and recoverable errors.

Unrecoverable errors:

 - You are using a library. Some function must be called in a specific way. The function checks whether this is done. It finds that the call is not correct. So there is obviously a bug in how the API is used. The function panics the program so that the programmer can fix it before deploying it to the user.
- unrecoverable errors are things like OOM, and basic things breaking in unforeseeable ways. Not being able to write to disk is not one of them, because that can happen for plenty of reasons. (Permission, disk is full, ...)

Recoverable errors:

 - A program works with user input. User input can be all kinds of wrong, so it must be checked whether for correctness and well-formedness. In case it isn't correct, you want to provide a proper, user-friendly error message that can be understood by just about anyone. Ideally with a nice backtrace for debugging purposes. That's what pacdef does and applies to your question.
- A web server uses worker threads. For some reason one of the worker threads dies. You may want to panic the server program, but you could also argue that the impact would be less severe if you start a new thread to replace the old one, and let the user request the webpage again. So you have a recoverable error.

Does that make sense?